// @antv/g6/lib/graph/controller/customGroup.js
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/*
 * @Author: moyee
 * @Date: 2019-07-30 12:10:26
 * @LastEditors: moyee
 * @LastEditTime: 2019-08-23 11:44:32
 * @Description: Group Controller
 */
var isString = require('@antv/util/lib/type/is-string');

var deepMix = require('@antv/util/lib/deep-mix');

var CustomGroup =
/*#__PURE__*/
function () {
  var _proto = CustomGroup.prototype;

  _proto.getDefaultCfg = function getDefaultCfg() {
    return {
      "default": {
        lineWidth: 1,
        stroke: '#A3B1BF',
        // lineDash: [ 5, 5 ],
        strokeOpacity: 0.9,
        fill: '#F3F9FF',
        fillOpacity: 0.8,
        opacity: 0.8,
        minDis: 20,
        maxDis: 40
      },
      hover: {
        stroke: '#faad14',
        fill: '#ffe58f',
        fillOpacity: 0.3,
        opacity: 0.3,
        lineWidth: 3
      },
      // 收起状态样式
      collapse: {
        r: 30,
        width: 80,
        height: 40,
        // lineDash: [ 5, 5 ],
        stroke: '#A3B1BF',
        lineWidth: 3,
        fill: '#F3F9FF'
      },
      icon: 'https://gw.alipayobjects.com/zos/rmsportal/MXXetJAxlqrbisIuZxDO.svg',
      title: {
        show: false,
        text: '新建群组',
        stroke: '#444',
        offsetX: 0,
        offsetY: 0
      },
      operatorBtn: {
        collapse: {
          img: 'https://gw.alipayobjects.com/zos/rmsportal/uZVdwjJGqDooqKLKtvGA.svg',
          width: 16,
          height: 16
        },
        expand: {
          width: 16,
          height: 16,
          img: 'https://gw.alipayobjects.com/zos/rmsportal/MXXetJAxlqrbisIuZxDO.svg'
        }
      },
      visible: false
    };
  };

  function CustomGroup(graph) {
    // const { cfg = {} } = options;
    this.graph = graph;
    window.graph = graph;
    var groupStyle = graph.get('groupStyle');
    this.styles = deepMix({}, this.getDefaultCfg(), groupStyle); // 创建的群组集合

    this.customGroup = {}; // 群组初始位置集合

    this.groupOriginBBox = {};
    this.delegateInGroup = {};
  }
  /**
   * 生成群组
   * @param {string} groupId 群组ID
   * @param {array} nodes 群组中的节点集合
   * @param {string} type 群组类型，默认为circle，支持rect
   * @param {number} zIndex 群组层级，默认为0
   * @param {boolean} updateDataModel 是否更新节点数据，默认为false，只有当手动创建group时才为true
   * @memberof ItemGroup
   */


  _proto.create = function create(groupId, nodes, type, zIndex, updateDataModel) {
    if (type === void 0) {
      type = 'circle';
    }

    if (zIndex === void 0) {
      zIndex = 0;
    }

    if (updateDataModel === void 0) {
      updateDataModel = false;
    }

    var graph = this.graph;
    var customGroup = graph.get('customGroup');
    var nodeGroup = customGroup.addGroup({
      id: groupId,
      zIndex: zIndex
    });
    var autoPaint = graph.get('autoPaint');
    graph.setAutoPaint(false);
    var _this$styles = this.styles,
        defaultStyle = _this$styles["default"],
        title = _this$styles.title; // 计算群组左上角左边、宽度、高度及x轴方向上的最大值

    var _this$calculationGrou = this.calculationGroupPosition(nodes),
        x = _this$calculationGrou.x,
        y = _this$calculationGrou.y,
        width = _this$calculationGrou.width,
        height = _this$calculationGrou.height,
        maxX = _this$calculationGrou.maxX;

    var paddingValue = this.getGroupPadding(groupId);
    var groupBBox = graph.get('groupBBoxs');
    groupBBox[groupId] = {
      x: x,
      y: y,
      width: width,
      height: height,
      maxX: maxX
    }; // step 1：绘制群组外框

    var keyShape = null;

    if (type === 'circle') {
      var r = width > height ? width / 2 : height / 2;
      var cx = (width + 2 * x) / 2;
      var cy = (height + 2 * y) / 2;
      keyShape = nodeGroup.addShape('circle', {
        attrs: _extends({}, defaultStyle, {
          x: cx,
          y: cy,
          r: r + nodes.length * 10 + paddingValue
        }),
        capture: true,
        zIndex: zIndex,
        groupId: groupId
      }); // 更新群组及属性样式

      this.setDeletageGroupByStyle(groupId, nodeGroup, {
        width: width,
        height: height,
        x: cx,
        y: cy,
        r: r
      });
    } else {
      keyShape = nodeGroup.addShape('rect', {
        attrs: _extends({}, defaultStyle, {
          x: x - paddingValue,
          y: y - paddingValue,
          width: width + nodes.length * 10 + paddingValue,
          height: height + nodes.length * 10 + paddingValue
        }),
        capture: true,
        zIndex: zIndex,
        groupId: groupId
      }); // 更新群组及属性样式

      this.setDeletageGroupByStyle(groupId, nodeGroup, {
        width: width,
        height: height,
        x: x,
        y: y,
        btnOffset: maxX - 3
      });
    } // 添加group标题


    if (title) {
      var show = title.show,
          offsetX = title.offsetX,
          offsetY = title.offsetY,
          titleStyle = _objectWithoutPropertiesLoose(title, ["show", "offsetX", "offsetY"]);

      if (show) {
        nodeGroup.addShape('text', {
          attrs: _extends({}, titleStyle, {
            x: x + offsetX,
            y: y + offsetY
          }),
          className: 'group-title'
        });
      }
    }

    nodeGroup.set('keyShape', keyShape); // 设置graph中groupNodes的值

    graph.get('groupNodes')[groupId] = nodes; // 只有手动创建group时执行以下逻辑

    if (updateDataModel) {
      // 如果是手动创建group，则原始数据中是没有groupId信息的，需要将groupId添加到node中
      nodes.forEach(function (nodeId) {
        var node = graph.findById(nodeId);
        var model = node.getModel();

        if (!model.groupId) {
          model.groupId = groupId;
        }
      });
    }

    this.setGroupOriginBBox(groupId, keyShape.getBBox());
    graph.setAutoPaint(autoPaint);
    graph.paint();
  }
  /**
   * 修改Group样式
   * @param {Item} keyShape 群组的keyShape
   * @param {Object | String} style 样式
   */
  ;

  _proto.setGroupStyle = function setGroupStyle(keyShape, style) {
    if (!keyShape || keyShape.get('destroyed')) {
      return;
    }

    var styles = {};
    var _this$styles2 = this.styles,
        hoverStyle = _this$styles2.hover,
        defaultStyle = _this$styles2["default"];

    if (isString(style)) {
      if (style === 'default') {
        styles = deepMix({}, defaultStyle);
      } else if (style === 'hover') {
        styles = deepMix({}, hoverStyle);
      }
    } else {
      styles = deepMix({}, defaultStyle, style);
    }

    for (var s in styles) {
      keyShape.attr(s, styles[s]);
    }
  }
  /**
   * 根据GroupID计算群组位置，包括左上角左边及宽度和高度
   *
   * @param {object} nodes 符合条件的node集合：选中的node或具有同一个groupID的node
   * @return {object} 根据节点计算出来的包围盒坐标
   * @memberof ItemGroup
   */
  ;

  _proto.calculationGroupPosition = function calculationGroupPosition(nodes) {
    var graph = this.graph;
    var minx = Infinity;
    var maxx = -Infinity;
    var miny = Infinity;
    var maxy = -Infinity; // 获取已节点的所有最大最小x y值

    for (var _iterator = nodes, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
      var _ref;

      if (_isArray) {
        if (_i >= _iterator.length) break;
        _ref = _iterator[_i++];
      } else {
        _i = _iterator.next();
        if (_i.done) break;
        _ref = _i.value;
      }

      var id = _ref;
      var element = isString(id) ? graph.findById(id) : id;
      var bbox = element.getBBox();
      var minX = bbox.minX,
          minY = bbox.minY,
          maxX = bbox.maxX,
          maxY = bbox.maxY;

      if (minX < minx) {
        minx = minX;
      }

      if (minY < miny) {
        miny = minY;
      }

      if (maxX > maxx) {
        maxx = maxX;
      }

      if (maxY > maxy) {
        maxy = maxY;
      }
    }

    var x = Math.floor(minx);
    var y = Math.floor(miny);
    var width = Math.ceil(maxx) - x;
    var height = Math.ceil(maxy) - y;
    return {
      x: x,
      y: y,
      width: width,
      height: height,
      maxX: Math.ceil(maxx)
    };
  }
  /**
   * 当group中含有group时，获取padding值
   * @param {string} groupId 节点分组ID
   * @return {number} 在x和y方向上的偏移值
   */
  ;

  _proto.getGroupPadding = function getGroupPadding(groupId) {
    var graph = this.graph;
    var defaultStyle = this.styles["default"]; // 检测操作的群组中是否包括子群组

    var groups = graph.get('groups');
    var hasSubGroup = !!groups.filter(function (g) {
      return g.parentId === groupId;
    }).length > 0;
    var paddingValue = hasSubGroup ? defaultStyle.maxDis : defaultStyle.minDis;
    return paddingValue;
  }
  /**
   * 设置群组初始位置的bbox，使用rect模拟
   *
   * @param {string} groupId 群组ID
   * @param {object} bbox 群组keyShape包围盒
   * @memberof ItemGroup
   */
  ;

  _proto.setGroupOriginBBox = function setGroupOriginBBox(groupId, bbox) {
    this.groupOriginBBox[groupId] = bbox;
  }
  /**
   * 获取群组初始位置及每次拖动后的位置
   *
   * @param {string} groupId 群组ID
   * @return {object} 指定groupId的原始BBox
   * @memberof ItemGroup
   */
  ;

  _proto.getGroupOriginBBox = function getGroupOriginBBox(groupId) {
    return this.groupOriginBBox[groupId];
  }
  /**
   * 设置群组对象及属性值
   *
   * @param {string} groupId 群组ID
   * @param {Group} deletage 群组元素
   * @param {object} property 属性值，里面包括width、height和maxX
   * @memberof ItemGroup
   */
  ;

  _proto.setDeletageGroupByStyle = function setDeletageGroupByStyle(groupId, deletage, property) {
    var width = property.width,
        height = property.height,
        x = property.x,
        y = property.y,
        r = property.r,
        btnOffset = property.btnOffset;
    var customGroupStyle = this.customGroup[groupId];

    if (!customGroupStyle) {
      // 首次赋值
      this.customGroup[groupId] = {
        nodeGroup: deletage,
        groupStyle: {
          width: width,
          height: height,
          x: x,
          y: y,
          r: r,
          btnOffset: btnOffset
        }
      };
    } else {
      // 更新时候merge配置项
      var groupStyle = customGroupStyle.groupStyle;
      var styles = deepMix({}, groupStyle, property);
      this.customGroup[groupId] = {
        nodeGroup: deletage,
        groupStyle: styles
      };
    }
  }
  /**
   * 根据群组ID获取群组及属性对象
   *
   * @param {string} groupId 群组ID
   * @return {Item} 群组
   * @memberof ItemGroup
   */
  ;

  _proto.getDeletageGroupById = function getDeletageGroupById(groupId) {
    return this.customGroup[groupId];
  }
  /**
   * 收起和展开群组
   * @param {string} groupId 群组ID
   */
  ;

  _proto.collapseExpandGroup = function collapseExpandGroup(groupId) {
    var customGroup = this.getDeletageGroupById(groupId);
    var nodeGroup = customGroup.nodeGroup;
    var hasHidden = nodeGroup.get('hasHidden'); // 该群组已经处于收起状态，需要展开

    if (hasHidden) {
      nodeGroup.set('hasHidden', false);
      this.expandGroup(groupId);
    } else {
      nodeGroup.set('hasHidden', true);
      this.collapseGroup(groupId);
    }
  }
  /**
   * 将临时节点递归地设置到groupId及父节点上
   * @param {string} groupId 群组ID
   * @param {string} tmpNodeId 临时节点ID
   */
  ;

  _proto.setGroupTmpNode = function setGroupTmpNode(groupId, tmpNodeId) {
    var graph = this.graph;
    var graphNodes = graph.get('groupNodes');
    var groups = graph.get('groups');

    if (graphNodes[groupId].indexOf(tmpNodeId) < 0) {
      graphNodes[groupId].push(tmpNodeId);
    } // 获取groupId的父群组


    var parentGroup = groups.filter(function (g) {
      return g.id === groupId;
    });
    var parentId = null;

    if (parentGroup.length > 0) {
      parentId = parentGroup[0].parentId;
    } // 如果存在父群组，则把临时元素也添加到父群组中


    if (parentId) {
      this.setGroupTmpNode(parentId, tmpNodeId);
    }
  }
  /**
   * 收起群组，隐藏群组中的节点及边，群组外部相邻的边都连接到群组上
   *
   * @param {string} id 群组ID
   * @memberof ItemGroup
   */
  ;

  _proto.collapseGroup = function collapseGroup(id) {
    var _this = this;

    var self = this;
    var customGroup = this.getDeletageGroupById(id);
    var nodeGroup = customGroup.nodeGroup; // 收起群组后的默认样式

    var collapse = this.styles.collapse;
    var graph = this.graph;
    var groupType = graph.get('groupType');
    var autoPaint = graph.get('autoPaint');
    graph.setAutoPaint(false);
    var nodesInGroup = graph.get('groupNodes')[id];

    var _this$calculationGrou2 = this.calculationGroupPosition(nodesInGroup),
        w = _this$calculationGrou2.width,
        h = _this$calculationGrou2.height; // 更新Group的大小


    var keyShape = nodeGroup.get('keyShape');

    var r = collapse.r,
        width = collapse.width,
        height = collapse.height,
        otherStyle = _objectWithoutPropertiesLoose(collapse, ["r", "width", "height"]);

    for (var style in otherStyle) {
      keyShape.attr(style, otherStyle[style]);
    }

    var options = {
      groupId: id,
      id: id + "-custom-node",
      x: keyShape.attr('x'),
      y: keyShape.attr('y'),
      style: {
        r: r
      },
      shape: 'circle'
    }; // 收起群组时候动画

    if (groupType === 'circle') {
      var radius = w > h ? w / 2 : h / 2; // const radius = wh + nodesInGroup.length * 10

      keyShape.animate({
        onFrame: function onFrame(ratio) {
          if (ratio === 1) {
            self.setGroupOriginBBox(id, keyShape.getBBox());
          }

          return {
            r: radius - ratio * (radius - r)
          };
        }
      }, 500, 'easeCubic');
    } else if (groupType === 'rect') {
      keyShape.animate({
        onFrame: function onFrame(ratio) {
          if (ratio === 1) {
            self.setGroupOriginBBox(id, keyShape.getBBox());
          }

          return {
            width: w - ratio * (w - width),
            height: h - ratio * (h - height)
          };
        }
      }, 500, 'easeCubic');
      options = {
        groupId: id,
        id: id + "-custom-node",
        x: keyShape.attr('x') + width / 2,
        y: keyShape.attr('y') + height / 2,
        size: [width, height],
        shape: 'rect'
      };
    }

    var edges = graph.getEdges(); // 获取所有source在群组外，target在群组内的边

    var sourceOutTargetInEdges = edges.filter(function (edge) {
      var model = edge.getModel();
      return !nodesInGroup.includes(model.source) && nodesInGroup.includes(model.target);
    }); // 获取所有source在群组外，target在群组内的边

    var sourceInTargetOutEdges = edges.filter(function (edge) {
      var model = edge.getModel();
      return nodesInGroup.includes(model.source) && !nodesInGroup.includes(model.target);
    }); // 获取群组中节点之间的所有边

    var edgeAllInGroup = edges.filter(function (edge) {
      var model = edge.getModel();
      return nodesInGroup.includes(model.source) && nodesInGroup.includes(model.target);
    }); // 隐藏群组中的所有节点

    nodesInGroup.forEach(function (nodeId) {
      var node = graph.findById(nodeId);
      var model = node.getModel();
      var groupId = model.groupId;

      if (groupId && groupId !== id) {
        // 存在群组，则隐藏
        var currentGroup = _this.getDeletageGroupById(groupId);

        var _nodeGroup = currentGroup.nodeGroup;

        _nodeGroup.hide();
      }

      node.hide();
    });
    edgeAllInGroup.forEach(function (edge) {
      var source = edge.getSource();
      var target = edge.getTarget();

      if (source.isVisible() && target.isVisible()) {
        edge.show();
      } else {
        edge.hide();
      }
    }); // 群组中存在source和target其中有一个在群组内，一个在群组外的情况

    if (sourceOutTargetInEdges.length > 0 || sourceInTargetOutEdges.length > 0) {
      var delegateNode = graph.add('node', options);
      delegateNode.set('capture', false);
      delegateNode.hide();
      this.delegateInGroup[id] = {
        delegateNode: delegateNode
      }; // 将临时添加的节点加入到群组中，以便拖动节点时候线跟着拖动

      this.setGroupTmpNode(id, id + "-custom-node");
      this.updateEdgeInGroupLinks(id, sourceOutTargetInEdges, sourceInTargetOutEdges);
    }

    graph.paint();
    graph.setAutoPaint(autoPaint);
  }
  /**
   * 收起群组时生成临时的节点，用于连接群组外的节点
   *
   * @param {string} groupId 群组ID
   * @param {array} sourceOutTargetInEdges 出度的边
   * @param {array} sourceInTargetOutEdges 入度的边
   * @memberof ItemGroup
   */
  ;

  _proto.updateEdgeInGroupLinks = function updateEdgeInGroupLinks(groupId, sourceOutTargetInEdges, sourceInTargetOutEdges) {
    var graph = this.graph; // 更新source在外的节点

    var edgesOuts = {};
    sourceOutTargetInEdges.map(function (edge) {
      var model = edge.getModel();
      var id = edge.get('id');
      var target = model.target;
      edgesOuts[id] = target;
      graph.updateItem(edge, {
        target: groupId + "-custom-node"
      });
      return true;
    }); // 更新target在外的节点

    var edgesIn = {};
    sourceInTargetOutEdges.map(function (edge) {
      var model = edge.getModel();
      var id = edge.get('id');
      var source = model.source;
      edgesIn[id] = source;
      graph.updateItem(edge, {
        source: groupId + "-custom-node"
      });
      return true;
    }); // 缓存群组groupId下的edge和临时生成的node节点

    this.delegateInGroup[groupId] = deepMix({
      sourceOutTargetInEdges: sourceOutTargetInEdges,
      sourceInTargetOutEdges: sourceInTargetOutEdges,
      edgesOuts: edgesOuts,
      edgesIn: edgesIn
    }, this.delegateInGroup[groupId]);
  }
  /**
   * 展开群组，恢复群组中的节点及边
   *
   * @param {string} id 群组ID
   * @memberof ItemGroup
   */
  ;

  _proto.expandGroup = function expandGroup(id) {
    var _this2 = this;

    var graph = this.graph;
    var groupType = graph.get('groupType');
    var self = this;
    var autoPaint = graph.get('autoPaint');
    graph.setAutoPaint(false); // 显示之前隐藏的节点和群组

    var nodesInGroup = graph.get('groupNodes')[id];

    var _this$calculationGrou3 = this.calculationGroupPosition(nodesInGroup),
        width = _this$calculationGrou3.width,
        height = _this$calculationGrou3.height; // const noCustomNodes = nodesInGroup.filter(node => node.indexOf('custom-node') === -1);


    var _this$getDeletageGrou = this.getDeletageGroupById(id),
        nodeGroup = _this$getDeletageGrou.nodeGroup;

    var keyShape = nodeGroup.get('keyShape');
    var _this$styles3 = this.styles,
        defaultStyle = _this$styles3["default"],
        collapse = _this$styles3.collapse;

    for (var style in defaultStyle) {
      keyShape.attr(style, defaultStyle[style]);
    } // 检测操作的群组中是否包括子群组


    var paddingValue = this.getGroupPadding(id);

    if (groupType === 'circle') {
      var r = width > height ? width / 2 : height / 2;
      keyShape.animate({
        onFrame: function onFrame(ratio) {
          if (ratio === 1) {
            self.setGroupOriginBBox(id, keyShape.getBBox());
          }

          return {
            r: collapse.r + ratio * (r + nodesInGroup.length * 10 - collapse.r + paddingValue)
          };
        }
      }, 500, 'easeCubic');
    } else if (groupType === 'rect') {
      var w = collapse.width,
          h = collapse.height;
      keyShape.animate({
        onFrame: function onFrame(ratio) {
          if (ratio === 1) {
            self.setGroupOriginBBox(id, keyShape.getBBox());
          }

          return {
            width: w + ratio * (width - w + paddingValue),
            height: h + ratio * (height - h + paddingValue)
          };
        }
      }, 500, 'easeCubic');
    } // 群组动画一会后再显示节点和边


    setTimeout(function () {
      nodesInGroup.forEach(function (nodeId) {
        var node = graph.findById(nodeId);
        var model = node.getModel();
        var groupId = model.groupId;

        if (groupId && groupId !== id) {
          // 存在群组，则显示
          var currentGroup = _this2.getDeletageGroupById(groupId);

          var _nodeGroup2 = currentGroup.nodeGroup;

          _nodeGroup2.show();

          var hasHidden = _nodeGroup2.get('hasHidden');

          if (!hasHidden) {
            node.show();
          }
        } else {
          node.show();
        }
      });
      var edges = graph.getEdges(); // 获取群组中节点之间的所有边

      var edgeAllInGroup = edges.filter(function (edge) {
        var model = edge.getModel();
        return nodesInGroup.includes(model.source) || nodesInGroup.includes(model.target);
      });
      edgeAllInGroup.forEach(function (edge) {
        var source = edge.getSource();
        var target = edge.getTarget();

        if (source.isVisible() && target.isVisible()) {
          edge.show();
        }
      });
    }, 300);
    var delegates = this.delegateInGroup[id];

    if (delegates) {
      var sourceOutTargetInEdges = delegates.sourceOutTargetInEdges,
          sourceInTargetOutEdges = delegates.sourceInTargetOutEdges,
          edgesOuts = delegates.edgesOuts,
          edgesIn = delegates.edgesIn,
          delegateNode = delegates.delegateNode; // 恢复source在外的节点

      sourceOutTargetInEdges.map(function (edge) {
        var id = edge.get('id');
        var sourceOuts = edgesOuts[id];
        graph.updateItem(edge, {
          target: sourceOuts
        });
        return true;
      }); // 恢复target在外的节点

      sourceInTargetOutEdges.map(function (edge) {
        var id = edge.get('id');
        var sourceIn = edgesIn[id];
        graph.updateItem(edge, {
          source: sourceIn
        });
        return true;
      }); // 删除群组中的临时节点ID

      var tmpNodeModel = delegateNode.getModel();
      this.deleteTmpNode(id, tmpNodeModel.id);
      graph.remove(delegateNode);
      delete this.delegateInGroup[id];
    }

    graph.setAutoPaint(autoPaint);
    graph.paint();
  };

  _proto.deleteTmpNode = function deleteTmpNode(groupId, tmpNodeId) {
    var graph = this.graph;
    var groups = graph.get('groups');
    var nodesInGroup = graph.get('groupNodes')[groupId];
    var index = nodesInGroup.indexOf(tmpNodeId);
    nodesInGroup.splice(index, 1); // 获取groupId的父群组

    var parentGroup = groups.filter(function (g) {
      return g.id === groupId;
    });
    var parentId = null;

    if (parentGroup.length > 0) {
      parentId = parentGroup[0].parentId;
    } // 如果存在父群组，则把临时元素也添加到父群组中


    if (parentId) {
      this.deleteTmpNode(parentId, tmpNodeId);
    }
  }
  /**
   * 删除节点分组
   * @param {string} groupId 节点分组ID
   * @memberof ItemGroup
   */
  ;

  _proto.remove = function remove(groupId) {
    var graph = this.graph;
    var customGroup = this.getDeletageGroupById(groupId);

    if (!customGroup) {
      console.warn("\u8BF7\u786E\u8BA4\u8F93\u5165\u7684groupId " + groupId + " \u662F\u5426\u6709\u8BEF\uFF01");
      return;
    }

    var nodeGroup = customGroup.nodeGroup;
    var autoPaint = graph.get('autoPaint');
    graph.setAutoPaint(false);
    var groupNodes = graph.get('groupNodes');
    var nodes = groupNodes[groupId]; // 删除原群组中node中的groupID

    nodes.forEach(function (nodeId) {
      var node = graph.findById(nodeId);
      var model = node.getModel();
      var gId = model.groupId;

      if (!gId) {
        return;
      }

      if (groupId === gId) {
        delete model.groupId; // 使用没有groupID的数据更新节点

        graph.updateItem(node, model);
      }
    });
    nodeGroup.destroy(); // 删除customGroup中groupId的数据

    delete this.customGroup[groupId]; // 删除groups数据中的groupId

    var groups = graph.get('groups');

    if (groups.length > 0) {
      var filterGroup = groups.filter(function (group) {
        return group.id !== groupId;
      });
      graph.set('groups', filterGroup);
    }

    var parentGroupId = null;
    var parentGroupData = null;

    for (var _iterator2 = groups, _isArray2 = Array.isArray(_iterator2), _i2 = 0, _iterator2 = _isArray2 ? _iterator2 : _iterator2[Symbol.iterator]();;) {
      var _ref2;

      if (_isArray2) {
        if (_i2 >= _iterator2.length) break;
        _ref2 = _iterator2[_i2++];
      } else {
        _i2 = _iterator2.next();
        if (_i2.done) break;
        _ref2 = _i2.value;
      }

      var group = _ref2;

      if (groupId !== group.id) {
        continue;
      }

      parentGroupId = group.parentId;
      parentGroupData = group;
      break;
    }

    if (parentGroupData) {
      delete parentGroupData.parentId;
    } // 删除groupNodes中的groupId数据


    delete groupNodes[groupId];

    if (parentGroupId) {
      groupNodes[parentGroupId] = groupNodes[parentGroupId].filter(function (node) {
        return !nodes.includes(node);
      });
    }

    graph.setAutoPaint(autoPaint);
    graph.paint();
  };

  _proto.destroy = function destroy() {
    this.graph = null;
    this.styles = {};
    this.customGroup = {};
    this.groupOriginBBox = {};
    this.delegateInGroup = {};
  };

  return CustomGroup;
}();

module.exports = CustomGroup;
